### PR TITLE
[Automation] Refactoring Automation Resources

### PR DESCRIPTION
**Pull request by vrdmr**
Refactoring Automation Resources to prevent the duplication of Models in different definitions file and creating a single common definition file for definitions/parameters used across Automation.

### PR information
- [x] The title of the PR is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).
- [x] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [ ] If applicable, the PR references the bug/issue that it fixes.
- [x] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).

### Quality of Swagger
- [x] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [x] My spec meets the review criteria:
  - [x] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [x] The spec follows the guidelines described in the [Swagger checklist](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md) document.
  - [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.

### CODE REVIEW

File: specification/automation/resource-manager/Microsoft.Automation/stable/2015-10-31/definitions.json [hunk 1]
```diff
@@ -1170,148 +797,39 @@
...
-        },
-        "nextLink": {
+        "uri": {
           "type": "string",
-          "description": "Gets or sets the next link."
-        }
-      },
-      "description": "The response model for the list job operation."
-    },
-    "DscConfigurationAssociationProperty": {
```
Comment (jianghaolu) [inline specification/automation/resource-manager/Microsoft.Automation/stable/2015-10-31/definitions.json]: Remove this property invalidates https://github.com/Azure/azure-rest-api-specs/pull/3190/files#diff-09777627978972db514896bb1119bc44R269

File: specification/automation/resource-manager/Microsoft.Automation/stable/2015-10-31/account.json [hunk 2]
```diff
@@ -287,31 +251,29 @@
...
-            "name": "resourceGroupName",
-            "in": "path",
-            "required": true,
-            "type": "string",
-            "pattern": "^[-\\w\\._]+$",
-            "description": "The resource group name."
+            "$ref": "../../common/v1/definitions.json#/parameters/ResourceGroupNameParameter"
+          },
+          {
+            "$ref": "../../common/v1/definitions.json#/parameters/AutomationAccountNameParameter"
```
Comment (jianghaolu) [inline specification/automation/resource-manager/Microsoft.Automation/stable/2015-10-31/account.json]: There's no automationAccountName needed here.

**Review by jianghaolu — CHANGES_REQUESTED** [review]
Please fix the semantic errors since they are preventing further validations to be run.

File: specification/automation/resource-manager/Microsoft.Automation/stable/2015-10-31/schedule.json [hunk 3]
```diff
@@ -331,6 +311,282 @@
...
+      "properties": {
+        "properties": {
+          "$ref": "#/definitions/ScheduleProperties",
+          "x-ms-client-flatten": true,
+          "description": "Gets or sets the properties of the schedule."
+        }
+      },
+      "allOf": [
+        {
+          "$ref": "../../common/v1/definitions.json#/definitions/ProxyResource"
```
Comment (vrdmr) [inline specification/automation/resource-manager/Microsoft.Automation/stable/2015-10-31/schedule.json]: @DeveloperTommy Added Schedule as a ProxyResource.

File: specification/automation/resource-manager/Microsoft.Automation/stable/2015-10-31/schedule.json [hunk 4]
```diff
@@ -74,22 +70,22 @@
...
+            "$ref": "../../common/v1/definitions.json#/parameters/SubscriptionIdParameter"
           },
           {
-            "$ref": "./definitions.json#/parameters/ApiVersionParameter"
+            "$ref": "../../common/v1/definitions.json#/parameters/ApiVersionParameter"
           }
         ],
         "responses": {
-          "201": {
+          "200": {
```
Comment (vrdmr) [inline specification/automation/resource-manager/Microsoft.Automation/stable/2015-10-31/schedule.json]: Changed the 201 to 200.

File: specification/automation/resource-manager/Microsoft.Automation/stable/2015-10-31/examples/createOrUpdateSchedule.json [hunk 5]
```diff
@@ -18,7 +18,7 @@
     }
   },
   "responses": {
-    "201": {
+    "200": {
```
Comment (vrdmr) [inline specification/automation/resource-manager/Microsoft.Automation/stable/2015-10-31/examples/createOrUpdateSchedule.json]: @DeveloperTommy Changed 201 to 200.

**Comment by vrdmr** [comment]
@jianghaolu Could you please take a look at this PR. Thanks.

File: specification/automation/resource-manager/Microsoft.Automation/stable/2015-10-31/definitions.json (hunk 1 above)
Comment (vrdmr) [reply specification/automation/resource-manager/Microsoft.Automation/stable/2015-10-31/definitions.json]: Fixed.

**Review by jianghaolu — CHANGES_REQUESTED** [review]

File: specification/automation/resource-manager/Microsoft.Automation/preview/2017-05-15-preview/softwareUpdateConfiguration.json [hunk 6]
```diff
@@ -58,54 +58,54 @@
...
             }
           },
           "201": {
             "description": "Software update configuration is created.",
             "schema": {
-              "$ref": "./definitions.json#/definitions/softwareUpdateConfiguration"
+              "$ref": "#/definitions/softwareUpdateConfiguration"
             }
           },
           "default": {
```
[the remark this answers is not in the document]
Comment (vrdmr) [reply specification/automation/resource-manager/Microsoft.Automation/preview/2017-05-15-preview/softwareUpdateConfiguration.json]: Fixed.

File: specification/automation/resource-manager/Microsoft.Automation/preview/2017-05-15-preview/softwareUpdateConfiguration.json [hunk 7]
```diff
@@ -214,7 +214,7 @@
...
```
[the remark this answers is not in the document]
Comment (vrdmr) [reply specification/automation/resource-manager/Microsoft.Automation/preview/2017-05-15-preview/softwareUpdateConfiguration.json]: Fixed.

File: specification/automation/resource-manager/Microsoft.Automation/stable/2015-10-31/credential.json [hunk 8]
```diff
@@ -53,14 +53,10 @@
...
           {
-            "name": "automationAccountName",
-            "in": "path",
-            "required": true,
-            "type": "string",
-            "description": "The automation account name."
+            "$ref": "../../common/v1/definitions.json#/parameters/AutomationAccountNameParameter"
           },
           {
             "name": "credentialName",
```
[the remark this answers is not in the document]
Comment (vrdmr) [reply specification/automation/resource-manager/Microsoft.Automation/stable/2015-10-31/credential.json]: Fixed.

File: specification/automation/resource-manager/Microsoft.Automation/stable/2015-10-31/job.json [hunk 9]
```diff
@@ -736,24 +696,19 @@
...
+        "Succeeded",	
+        "Suspended",	
+        "Processing"	
+      ],	
+      "x-ms-enum": {	
+        "name": "JobProvisioningState",	
+        "modelAsString": true	
+      }
     },
     "JobCreateParameters": {
```
[the remark this answers is not in the document]
Comment (vrdmr) [reply specification/automation/resource-manager/Microsoft.Automation/stable/2015-10-31/job.json]: @jianghaolu Its a proxy resource and would not have "Location", "Name" and "Tags".

File: specification/automation/resource-manager/Microsoft.Automation/stable/2015-10-31/job.json [hunk 10]
```diff
@@ -401,10 +373,10 @@
...
           {
-            "$ref": "./definitions.json#/parameters/SubscriptionIdParameter"
+            "$ref": "../../common/v1/definitions.json#/parameters/SubscriptionIdParameter"
           },
           {
-            "$ref": "./definitions.json#/parameters/ApiVersionParameter"
+            "$ref": "../../common/v1/definitions.json#/parameters/ApiVersionParameter"
           }
         ],
         "responses": {
```
[the remark this answers is not in the document]
Comment (vrdmr) [reply specification/automation/resource-manager/Microsoft.Automation/stable/2015-10-31/job.json]: I've fixed this. But we have a new version of the Jobs resource `2017-05-15-preview` and we have updated the SDK to use the new version as well.

File: specification/automation/resource-manager/Microsoft.Automation/stable/2015-10-31/account.json [hunk 11]
```diff
@@ -520,48 +461,408 @@
...
+            "$ref": "#/definitions/Usage"
+          },
+          "description": "Gets or sets usage."
+        }
+      },
+      "description": "The response model for the get usage operation."
+    },
+    "Key": {
+      "properties": {
+        "keyName": {
```
[the remark this answers is not in the document]
Comment (vrdmr) [reply specification/automation/resource-manager/Microsoft.Automation/stable/2015-10-31/account.json]: Hi @jianghaolu - Our service returns the properties in Uppercase, and if I define the property in upper case, I get the following error - 
```ERROR (DefinitionsPropertiesNamesCamelCase/R3016/RPCViolation): Property named: 'KeyName', for definition: 'Key' must follow camelCase style. Example: 'keyName'. ```

Is it a blocking issue or something which can leave it as it is in this API version and we can fix in the newer API version? Please let me know.

File: specification/automation/resource-manager/Microsoft.Automation/preview/2017-05-15-preview/softwareUpdateConfiguration.json [hunk 12]
```diff
@@ -126,33 +126,33 @@
...
         ],
         "responses": {
           "200": {
             "description": "A single software update configuration.",
             "schema": {
-              "$ref": "./definitions.json#/definitions/softwareUpdateConfiguration"
+              "$ref": "#/definitions/softwareUpdateConfiguration"
             }
           },
           "404": {
```
Comment (jianghaolu) [inline specification/automation/resource-manager/Microsoft.Automation/preview/2017-05-15-preview/softwareUpdateConfiguration.json]: status codes "404" for operation "SoftwareUpdateConfigurations_GetByName" were present in the swagger spec, however they were not present in x-ms-examples.

File: specification/automation/resource-manager/Microsoft.Automation/preview/2017-05-15-preview/softwareUpdateConfiguration.json [hunk 13]
```diff
@@ -266,7 +266,7 @@
           "200": {
             "description": "Return list of software update configurations.",
             "schema": {
-              "$ref": "./definitions.json#/definitions/softwareUpdateConfigurationListResult"
+              "$ref": "#/definitions/softwareUpdateConfigurationListResult"
             }
           },
           "404": {
```
Comment (jianghaolu) [inline specification/automation/resource-manager/Microsoft.Automation/preview/2017-05-15-preview/softwareUpdateConfiguration.json]: Same here

File: specification/automation/resource-manager/Microsoft.Automation/stable/2015-10-31/account.json (hunk 11 above)
[the remark this answers is not in the document]
Comment (jianghaolu) [reply specification/automation/resource-manager/Microsoft.Automation/stable/2015-10-31/account.json]: That error is simply a style error - but if your swagger is not corresponding to the actual service then developers won't be able to use the clients generated from this.

[the remark this answers is not in the document]
Comment (jianghaolu) [reply specification/automation/resource-manager/Microsoft.Automation/stable/2015-10-31/account.json]: You might want to follow the suppression process to suppress this R3016 error: https://github.com/Azure/adx-documentation-pr/wiki/Swagger-Validation-Errors-Suppression

File: specification/automation/resource-manager/Microsoft.Automation/stable/2015-10-31/job.json (hunk 9 above)
[the remark this answers is not in the document]
Comment (jianghaolu) [reply specification/automation/resource-manager/Microsoft.Automation/stable/2015-10-31/job.json]: But in the example I linked above there is - you might want to double check if it's a proxy resource.

File: specification/automation/resource-manager/Microsoft.Automation/stable/2018-01-15/dscNodeConfiguration.json [hunk 14]
```diff
@@ -443,7 +443,7 @@
           "$ref": "#/definitions/DscConfigurationAssociationProperty",
           "description": "Gets or sets the configuration of the node."
         },
-        "incrementNodeConfigurationBuild": {
+        "IncrementNodeConfigurationBuild": {
```
Comment (jianghaolu) [inline specification/automation/resource-manager/Microsoft.Automation/stable/2018-01-15/dscNodeConfiguration.json]: Please update the corresponding fields in this example too: https://github.com/vrdmr/azure-rest-api-specs/blob/0489a8a2c0bf758e9593a7d21ff507e0899ff698/specification/automation/resource-manager/Microsoft.Automation/stable/2018-01-15/examples/createOrUpdateDscNodeConfiguration.json

Comment (vrdmr) [reply specification/automation/resource-manager/Microsoft.Automation/stable/2018-01-15/dscNodeConfiguration.json]: Fixed.

File: specification/automation/resource-manager/Microsoft.Automation/preview/2017-05-15-preview/softwareUpdateConfiguration.json (hunk 12 above)
Comment (vrdmr) [reply specification/automation/resource-manager/Microsoft.Automation/preview/2017-05-15-preview/softwareUpdateConfiguration.json]: Fixed.

File: specification/automation/resource-manager/Microsoft.Automation/preview/2017-05-15-preview/softwareUpdateConfiguration.json (hunk 13 above)
Comment (vrdmr) [reply specification/automation/resource-manager/Microsoft.Automation/preview/2017-05-15-preview/softwareUpdateConfiguration.json]: Fixed.

File: specification/automation/resource-manager/Microsoft.Automation/stable/2015-10-31/runbook.json [hunk 15]
```diff
@@ -244,31 +232,27 @@
...
-          },          
+          },
           {
-            "$ref": "./definitions.json#/parameters/ApiVersionParameter"
+            "$ref": "../../common/v1/definitions.json#/parameters/ApiVersionParameter"
           }
         ],
         "responses": {
-          "200": {            
+          "200": {
```
[the remark this answers is not in the document]
Comment (vrdmr) [reply specification/automation/resource-manager/Microsoft.Automation/stable/2015-10-31/runbook.json]: Same as above: It never generates 200. This was added to fix the linting issues in #2687.

File: specification/automation/resource-manager/Microsoft.Automation/stable/2015-10-31/runbook.json [hunk 16]
```diff
@@ -142,11 +134,11 @@
             "description": "The runbook draft content."
           },
           {
-            "$ref": "./definitions.json#/parameters/ApiVersionParameter"
+            "$ref": "../../common/v1/definitions.json#/parameters/ApiVersionParameter"
           }
         ],
         "responses": {
-          "200": {            
+          "200": {
```
[the remark this answers is not in the document]
Comment (vrdmr) [reply specification/automation/resource-manager/Microsoft.Automation/stable/2015-10-31/runbook.json]: It never generates 200. This was added to fix the linting issues in #2687.

File: specification/automation/resource-manager/Microsoft.Automation/stable/2015-10-31/account.json (hunk 11 above)
[the remark this answers is not in the document]
Comment (vrdmr) [reply specification/automation/resource-manager/Microsoft.Automation/stable/2015-10-31/account.json]: Fixed.

File: specification/automation/resource-manager/Microsoft.Automation/stable/2015-10-31/job.json (hunk 9 above)
[the remark this answers is not in the document]
Comment (vrdmr) [reply specification/automation/resource-manager/Microsoft.Automation/stable/2015-10-31/job.json]: I'll verify and confirm tomorrow.

File: specification/automation/resource-manager/Microsoft.Automation/stable/2015-10-31/job.json [hunk 17]
```diff
@@ -53,14 +53,10 @@
         },
```
[the remark this answers is not in the document]
Comment (vrdmr) [reply specification/automation/resource-manager/Microsoft.Automation/stable/2015-10-31/job.json]: We need to discuss this issue. Can we sync offline tomorrow?

**Comment by jianghaolu** [comment]
@vrdmr Sorry for the delay - Did you get a chance to confirm this? https://github.com/Azure/azure-rest-api-specs/pull/3190/files/0489a8a2c0bf758e9593a7d21ff507e0899ff698#diff-91d25383f8afa4ecf50a11d676f50454

**Comment by jianghaolu** [comment]
There are a few other errors reported in the validator which I'm trying to understand.

**Comment by jianghaolu** [comment]
It seems like one error that the model validator is reporting is related to all the `"advancedSchedule": null` in the examples. I tested it locally and it seems like removing those will resolve the errors from the model validator.

File: specification/automation/resource-manager/Microsoft.Automation/stable/2015-10-31/job.json (hunk 9 above)
[the remark this answers is not in the document]
Comment (vrdmr) [reply specification/automation/resource-manager/Microsoft.Automation/stable/2015-10-31/job.json]: Confirmed and removed.

File: specification/automation/resource-manager/Microsoft.Automation/stable/2015-10-31/runbook.json (hunk 16 above)
[the remark this answers is not in the document]
Comment (jianghaolu) [reply specification/automation/resource-manager/Microsoft.Automation/stable/2015-10-31/runbook.json]: If your service doesn't generate 200 you shouldn't put it in the spec. If the linter reports error then we should fix the linter.

[the remark this answers is not in the document]
Comment (vrdmr) [reply specification/automation/resource-manager/Microsoft.Automation/stable/2015-10-31/runbook.json]: Fixed with a suppression at for `LongRunningResponseStatusCode` in `readme.md` for this.

File: specification/automation/resource-manager/Microsoft.Automation/stable/2015-10-31/schedule.json [hunk 18]
```diff
@@ -247,9 +247,6 @@
           "200": {
             "description": "OK"
           },
-          "404": {
```
Comment (DeveloperTommy) [inline specification/automation/resource-manager/Microsoft.Automation/stable/2015-10-31/schedule.json]: I'm confused about why we are removing 404 here. If we try to get a schedule, shouldn't we return a 404 if it doesn't exist?

File: specification/automation/resource-manager/readme.md [hunk 19]
```diff
@@ -131,11 +128,23 @@ input-file:
...
 directive:
   - suppress: RequiredPropertiesMissingInResourceModel
-    from: definitions.json
+    from: runbook.json
     where: $.definitions.TestJob
   - suppress: BodyTopLevelProperties
-    from: definitions.json
+    from: runbook.json
     where: $.definitions.TestJob.properties
+  - suppress: DefinitionsPropertiesNamesCamelCase
```
Comment (balukambala) [inline specification/automation/resource-manager/readme.md]: signing off on this change

**Review by balukambala — APPROVED** [review]
Signing off on suppression changes on camel case

File: specification/automation/resource-manager/Microsoft.Automation/stable/2015-10-31/schedule.json (hunk 18 above)
Comment (vrdmr) [reply specification/automation/resource-manager/Microsoft.Automation/stable/2015-10-31/schedule.json]: In the specs, we only specify success cases and not the failure cases. If we start specifying error cases, the SDK does not throw an exception and treats it as a normal case.

For more info, @finiteattractor can describe it in more details.

Comment (DeveloperTommy) [reply specification/automation/resource-manager/Microsoft.Automation/stable/2015-10-31/schedule.json]: Thank you! Signing off.

**Review by DeveloperTommy — APPROVED** [review]

**Review by finiteattractor — APPROVED** [review]
Thanks Varad for doing this big refactoring! Changes look good to me.

**Review by jianghaolu — APPROVED** [review]

**Comment by Francisco-Gamino** [comment]
Thanks @jianghaolu!

**Pull request merged**